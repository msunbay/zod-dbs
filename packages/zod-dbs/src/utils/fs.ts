import { existsSync, readdirSync, unlinkSync } from 'node:fs';
import { mkdir } from 'node:fs/promises';
import path from 'node:path';
import { logDebug } from 'zod-dbs-core';

import type { ZodDbsTableType } from 'zod-dbs-core';

/**
 * Deletes all .ts files in the output schemas folder.
 */
export function clearTablesDirectory(outputPath: string) {
  // Check if the directory exists first
  if (!existsSync(outputPath)) {
    logDebug(`Directory ${outputPath} does not exist, nothing to clear`);
    return;
  }

  const files = readdirSync(outputPath, {
    recursive: true,
    withFileTypes: true,
  });

  for (const file of files) {
    if (file.isDirectory()) continue;
    if (!file.name.endsWith('.ts')) continue;

    const filePath = path.join(file.path, file.name);
    unlinkSync(filePath);
  }

  if (files.length > 0) {
    logDebug(`Deleted all .ts files in ${outputPath}`);
  }
}

export async function ensureFolder(folderPath: string) {
  const exists = existsSync(folderPath);

  if (!exists) {
    await mkdir(folderPath, { recursive: true });
  }
}

export const getOutputFolder = (type: ZodDbsTableType): string => {
  switch (type) {
    case 'table':
      return 'tables';
    case 'materialized_view':
      return 'materialized_views';
    case 'view':
      return 'views';
    case 'foreign_table':
      return 'foreign_tables';
    default:
      return 'others';
  }
};
