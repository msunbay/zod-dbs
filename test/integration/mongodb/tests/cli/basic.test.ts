import fs from 'node:fs/promises';
import path from 'node:path';

import {
  executeCli,
  getOutputFiles,
  getProviderOutputDir,
} from '../../../utils/cli.js';
import { getProviderConnectionString } from '../../../utils/context.js';

it('CLI generates correct zod schemas with basic options', async () => {
  const connectionString = getProviderConnectionString('mongodb');
  const outputDir = getProviderOutputDir('mongodb', 'basic');

  await executeCli(
    `--provider mongodb --connection-string "${connectionString}" --direct-connection --replica-set rs0 --output-dir "${outputDir}" --silent --module-resolution esm`
  );

  const outputFiles = await getOutputFiles(outputDir);

  for (const file of outputFiles) {
    const content = await fs.readFile(file, 'utf8');
    expect(content).toMatchSnapshot(path.relative(outputDir, file));
  }
});
