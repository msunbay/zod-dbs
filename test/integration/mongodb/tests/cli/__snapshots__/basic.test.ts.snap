// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CLI generates correct zod schemas with basic options > constants.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export const TABLE_USERS = 'users';
"
`;

exports[`CLI generates correct zod schemas with basic options > tables/index.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export * from './users/index.js';
"
`;

exports[`CLI generates correct zod schemas with basic options > tables/users/UserProfileSchema.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

import { z } from 'zod';


/**
 * Base read schema for the "UserProfile" object.
 * Validates raw rows read from the database (no casing transforms applied yet).
 */
export const UserProfileBaseSchema = z.object({
     /**
      * dataType: string
      */
    displayName: z.string().optional(),
     /**
      * dataType: int
      */
    age: z.number().optional(),
});

/**
 * The base record type for the "UserProfile" object.
 * This type represents the raw database record without case transforms.
 */
export type UserProfileBaseRecord = z.output<typeof UserProfileBaseSchema>;

/**
 * Read transform for the "UserProfile" object.
 * Maps database fields to app case properties.
 */
export const transformUserProfileBaseRecord = (data: UserProfileBaseRecord): {
    displayName?: UserProfileBaseRecord['displayName'],
    age?: UserProfileBaseRecord['age'],
} => ({
    displayName: data.displayName,
    age: data.age,
});

/**
 * Read schema for the "UserProfile" object (after casing transform).
 */
export const UserProfileSchema = UserProfileBaseSchema.transform(transformUserProfileBaseRecord);

type TableReadRecord = z.output<typeof UserProfileSchema>;

/**
 * Read record (casing transformed) for the "UserProfile" object.
 */
export interface UserProfileRecord {
    displayName?: TableReadRecord['displayName'];
    age?: TableReadRecord['age'];
}

"
`;

exports[`CLI generates correct zod schemas with basic options > tables/users/index.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export * from './schema.js';
"
`;

exports[`CLI generates correct zod schemas with basic options > tables/users/schema.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

import { z } from 'zod';
import { UserProfileSchema } from './UserProfileSchema.js';


/**
 * Base read schema for the "testdb.users" table.
 * Validates raw rows read from the database (no casing transforms applied yet).
 */
export const UsersTableBaseSchema = z.object({
     /**
      * dataType: objectid
      */
    _id: z.string(),
     /**
      * dataType: string
      */
    email: z.string(),
     /**
      * dataType: date
      */
    createdAt: z.coerce.date(),
     /**
      * dataType: object
      */
    profile: UserProfileSchema.nullable().transform((value) => value ?? undefined).optional(),
     /**
      * dataType: array
      */
    tags: z.array(z.any()).nullable().transform((value) => value ?? undefined).optional(),
});

/**
 * The base record type for the "testdb.users" table.
 * This type represents the raw database record without case transforms.
 */
export type UserBaseRecord = z.output<typeof UsersTableBaseSchema>;

/**
 * Read transform for the "testdb.users" table.
 * Maps database fields to app case properties.
 */
export const transformUserBaseRecord = (data: UserBaseRecord): {
    id: UserBaseRecord['_id'],
    email: UserBaseRecord['email'],
    createdAt: UserBaseRecord['createdAt'],
    profile?: UserBaseRecord['profile'],
    tags?: UserBaseRecord['tags'],
} => ({
    id: data._id,
    email: data.email,
    createdAt: data.createdAt,
    profile: data.profile,
    tags: data.tags,
});

/**
 * Read schema for the "testdb.users" table (after casing transform).
 */
export const UsersTableSchema = UsersTableBaseSchema.transform(transformUserBaseRecord);

/**
 * Base insert/write schema for the "testdb.users" table (no casing transforms).
 */
export const UsersTableInsertBaseSchema = z.object({
     /**
      * dataType: objectid
      */
    id: z.string(),
     /**
      * dataType: string
      */
    email: z.string(),
     /**
      * dataType: date
      */
    createdAt: z.date(),
     /**
      * dataType: object
      */
    profile: UserProfileSchema.nullable().optional(),
     /**
      * dataType: array
      */
    tags: z.array(z.any()).nullable().optional(),
});

/**
 * The base record type for the "testdb.users" table.
 * This type represents an insertable database record before casing transformations are applied.
 */
export type UserInsertBaseRecord = z.output<typeof UsersTableInsertBaseSchema>;

/**
 * The base record type for the "testdb.users" table.
 * This type represents an updatable database record before casing transformations are applied.
 */
export type UserUpdateBaseRecord = Partial<UserInsertBaseRecord>;

/**
 * Insert transform for the "testdb.users" table.
 * Maps app cased properties to database fields.
 */
export const transformUserInsertBaseRecord = (data: UserInsertBaseRecord): {
    _id: UserInsertBaseRecord['id'],
    email: UserInsertBaseRecord['email'],
    createdAt: UserInsertBaseRecord['createdAt'],
    profile?: UserInsertBaseRecord['profile'],
    tags?: UserInsertBaseRecord['tags'],
} => ({
    _id: data.id,
    email: data.email,
    createdAt: data.createdAt,
    profile: data.profile,
    tags: data.tags,
});

/**
 * Update transform for the "testdb.users" table.
 * Maps app cased properties to database fields.
 */
export const transformUserUpdateBaseRecord = (data: UserUpdateBaseRecord): {
    _id: UserUpdateBaseRecord['id'],
    email: UserUpdateBaseRecord['email'],
    createdAt: UserUpdateBaseRecord['createdAt'],
    profile?: UserUpdateBaseRecord['profile'],
    tags?: UserUpdateBaseRecord['tags'],
} => ({
    _id: data.id,
    email: data.email,
    createdAt: data.createdAt,
    profile: data.profile,
    tags: data.tags,
});

/**
 * Insert schema for the "testdb.users" table (after casing transform).
 */
export const UsersTableInsertSchema = UsersTableInsertBaseSchema.transform(transformUserInsertBaseRecord);

/**
 * Update schema for the "testdb.users" table (after casing transform).
 */
export const UsersTableUpdateSchema = UsersTableInsertBaseSchema.partial().transform(transformUserUpdateBaseRecord);

type TableInsertRecord = z.input<typeof UsersTableInsertSchema>;
type TableReadRecord = z.output<typeof UsersTableSchema>;

/**
 * Read record (casing transformed) for the "testdb.users" table.
 */
export interface UserRecord {
    id: TableReadRecord['id'];
    email: TableReadRecord['email'];
    createdAt: TableReadRecord['createdAt'];
    profile?: TableReadRecord['profile'];
    tags?: TableReadRecord['tags'];
}

/**
 * Insert record (casing transformed) for the "testdb.users" table.
 */
export interface UserInsertRecord {
    id: TableInsertRecord['id'];
    email: TableInsertRecord['email'];
    createdAt: TableInsertRecord['createdAt'];
    profile?: TableInsertRecord['profile'];
    tags?: TableInsertRecord['tags'];
}

/**
 * Updatable record (casing transformed) for the "testdb.users" table.
 */
export type UserUpdateRecord = Partial<UserInsertRecord>;

"
`;

exports[`CLI generates correct zod schemas with basic options > types.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export type Table = 
    | 'users'
"
`;
